### ISSUE TITLE
sender-callbacks 

### ISSUE THREAD

**Issue by monkeypants** [ausdigital/ausdigital-tap#28]
#7 made reference like:

> If error callback provided with submission, and signature does not pass validation, then the TAP MUST make error callback with submission ID and invalid signature error message.

This ticket is to expand on that idea, also with the idea that the message sender may want callbacks with notice of successful delivery or other state changes.

Note this ticket is complementary to ausdigital/ausdigital-tap-gw#8

**Comment by monkeypants** [comment]
I think this is fully covered in #33

**Comment by onthebreeze** [comment]
Just to clarify - we are talking about callbacks about the technical delivery of the message, not (for example) longer running business conversations.  As per https://github.com/ausdigital/ausdigital-tap/issues/19#issuecomment-298224157  So

* As a ledger sending an invoice, i'll use the TAP callbacks until I'm satisfied that the technical delivery is ok (ie signatures validate, routing exists, target system has got the document).  In most cases, that should all happen in a few seconds.
* But if the receiver of the invoice wants to send me a business level message response (eg "disputed" or "approved") then that will use the UBL business document response and is a seaprate asynchronous message int he conversation - probably many days or even weeks after the initial invoice is sent.  And that follows the business process model http://ausdigital.org/specs/ausdigital-bill/1.0/#state-lifecycle

With that distinction clear then yes, I think callbacks are a handy way to avoid having to hold open connections or to know whether to re-try.    A couple of questions come to mind:

* With reference to https://github.com/ausdigital/ausdigital-tap/issues/19#issuecomment-298224157, the transitions between "processing" and either "sent" or "error" will be pretty quick (few seconds) but the ones from "sent" to "delivered" or "error" could take a few days.  So, especially for the "stale" trigger, how does the receiving TAP know how long to wait before it comes to the conclusion that the ultimate recipient is never going to pick it up?
* What if the message sender is a mobile device or a system that is only occasionally online and may have no fixed callback URL?   Should it keep polling?  How frequently?

**Comment by monkeypants** [comment]
Yes, this is about message delivery (per #19), not business semantics of messages of protocols implemented with messages (e.g. ausdigital-bill).

I would hope it's mostly seconds, but there are lots of legitimate reasons to could take longer. For example, if a message is sent in the middle of a signature revocation, there is very little the TAP can do but wait for the new signature to propagate.

WRT "how long to wait" before declaring a message stale, that's the TAP's business. We might have ideas around that, but the TAP has given up trying to deliver when it says so. I would imagine a month might be typical.

WRT mobile devices or other simple/non-persistent message senders: I think the callback should be optional, if you don't want it then don't set it. Similarly:
 * yes, polling is an alternative to callback (needs to be clear in the spec)
 * the callback should be considered "best effort", there is no recourse if callback was unable to send.

So what's left to do here is define a specify callback protocol. It should have an explicitly narrow scope: error and success messages (WRT message delivery). We want to avoid overloading it with business semantics.

**Comment by onthebreeze** [comment]
OK that all sounds reasonable.  Although I'd suggest that a week rather than a month might be better guidance - accepting that it's a question for the receiver TAP.

**Comment by monkeypants** [comment]
Based on the TAP delivery state-chart in #19, the top-level states are:
 * error (terminal)
 * delivered (terminal)
 * sent
 * processing

There is no need for a "processing" callback, since by the time the POST receives a 200 response it's redundant. So assuming an initial state of "processing", the callbacks basically are sent, delivered and error. See #33 for detailed codes. 

per #18 ([this comment](https://github.com/ausdigital/ausdigital-tap/issues/18#issuecomment-305977776)), might want to include the compound conversation IDs so that the callback recipient can know the conversation ID (without necessarily knowing how to map the "id" to the message, i.e. if they didn't send it).

Note also, the number of callback subscribers is zero or more, so we will need to add to the TAP API a (potentially empty) list of callbacks, not a single value.
 
```json
{
    "data": {
        "type": "TapCallback",
        "id": "the message id provided with the 200 response when the message was initially posted",
        "attributes": {
            "messageHash": "value from the message.json file, i.e. unique for message content",
            "deliveryStatus": "one of 'sent', 'delivered' or 'error'",
            "deliveryMessage": "user-friendly message explaining the specific delivery status reason",
            "statusDate": "ISO 8601 datetime format with the timezone",
            "initiatorParticipantID": "scheme_X::id_A",
            "responderParticipantID": "scheme_Y::id_B",
            "initiatorReference": "foo-123456",
            "senderParticipantID": "scheme_X::id_A",
        }
    }
}
``` 

Notes:
 * the signing key of senderParticipantID is used to validate the message
 * statusDate is in the TAP's wall clock (and timezone), not necessarily sender's or recipient's.
 * deliveryMessage and deliveryStatus are as-per #33
  * initiatorParticipantID, responderParticipantID, initiatorReference and senderParticipantID are all as per #18 (compound conversation ID, responderParticipantID optional)

Also note: because anybody could send bogus callbacks to a callback URL, the callback should be a `Content-Type: multipart/form-data` (RFC2388) with a `callback` part (per above schema) and a detached `signature` part (same as the TAP message)

**Comment by monkeypants** [comment]
TODO:
 * add list of callback URLs to the message schema (i.e. posted to TAP)
 * update TAP spec to mention list of callback URLs
 * define callback API spec (you should implement this to receive callbacks)
 * define callback.schema
 * link TAP page to both message API and callback API (not a single API anymore)

**Comment by monkeypants** [comment]
Hmm, we should put error messages in the "error" top level list of errors (to be consistent with the message API). In the case of errors, deliveryMessage is ambiguous (therefor, should it be an optional field?)

**Comment by kshychko** [comment]
@monkeypants 
please review the PR #36 
once we merge the callback-swagger.yaml to master, I will update TAP page with additional raw for TAP callback API.

**Comment by monkeypants** [comment]
#36 is merged, do we still need to update the example in the spec?

**Comment by kshychko** [comment]
nope, the examples have been fixed.

**Comment by monkeypants** [comment]
@kshychko what is left to do on this ticket?
